src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp

```diff
@@ -1523,6 +1523,137 @@ void C2_MacroAssembler::vinsert(BasicType typ, XMMRegister dst, XMMRegister src,
   }
 }
 
+#ifdef _LP64
+void C2_MacroAssembler::vpgather8b_masked(BasicType elem_bt, XMMRegister dst, Register base, Register idx_base,
+                                          Register mask, Register midx, Register rtmp, int vlen_enc) {
+  vpxor(dst, dst, dst, vlen_enc);
+  if (elem_bt == T_SHORT) {
+    Label case0, case1, case2, case3;
+    Label* larr[] = { &case0, &case1, &case2, &case3 };
+    for (int i = 0; i < 4; i++) {
+      bt(mask, midx);
+      jccb(Assembler::carryClear, *larr[i]);
+      movl(rtmp, Address(idx_base, i*4));
+      pinsrw(dst, Address(base, rtmp, Address::times_2), i);
+      bind(*larr[i]);
+      incq(midx);
+    }
+  } else {
+    assert(elem_bt == T_BYTE, "");
+    Label case0, case1, case2, case3, case4, case5, case6, case7;
+    Label* larr[] = { &case0, &case1, &case2, &case3, &case4, &case5, &case6, &case7 };
+    for (int i = 0; i < 8; i++) {
+      bt(mask, midx);
+      jccb(Assembler::carryClear, *larr[i]);
+      movl(rtmp, Address(idx_base, i*4));
+      pinsrb(dst, Address(base, rtmp), i);
+      bind(*larr[i]);
+      incq(midx);
+    }
+  }
+}
+
+void C2_MacroAssembler::vpgather8b_masked_offset(BasicType elem_bt, XMMRegister dst, Register base, Register idx_base,
+                                                 Register offset, Register mask, Register midx, Register rtmp, int vlen_enc) {
+  vpxor(dst, dst, dst, vlen_enc);
+  if (elem_bt == T_SHORT) {
+    Label case0, case1, case2, case3;
+    Label* larr[] = { &case0, &case1, &case2, &case3 };
+    for (int i = 0; i < 4; i++) {
+      bt(mask, midx);
+      jccb(Assembler::carryClear, *larr[i]);
+      movl(rtmp, Address(idx_base, i*4));
+      addl(rtmp, offset);
+      pinsrw(dst, Address(base, rtmp, Address::times_2), i);
+      bind(*larr[i]);
+      incq(midx);
+    }
+  } else {
+    assert(elem_bt == T_BYTE, "");
+    Label case0, case1, case2, case3, case4, case5, case6, case7;
+    Label* larr[] = { &case0, &case1, &case2, &case3, &case4, &case5, &case6, &case7 };
+    for (int i = 0; i < 8; i++) {
+      bt(mask, midx);
```

Comment:
Could we not use smaller length bt and inc instructions (e.g. 32 bit one) here as we know that we dont need 64 bits of mask here? That way we will have smaller instruction encoding.

Reply:
I get your point it may save prefix byte for short vectors in one case, but REX2 may not be avoidable if allocator picks a register from higher register bank (r8-15), mask corresponding to Byte64 does need 64 bits.